validmind/unit_metrics/composite.py

```diff
@@ -0,0 +1,243 @@
+
+            self.generic_visit(node)
+
+        def visit_Call(self, node):
+            if isinstance(node.func, ast.Attribute) and isinstance(
+                node.func.value, ast.Name
+            ):
+                if node.func.value.id == "self" and node.func.attr in methods:
+                    method_name = node.func.attr
+
+                    if method_name not in self.visited_methods:
+                        self.visited_methods.add(method_name)
+                        self.visit(methods[method_name])
+
+            self.generic_visit(node)
+
+    visitor = Visitor()
+    visitor.visit(methods["run"])
+
+    return visitor.properties
+
+
+class MetricProtocol(Protocol):
+    unit_metrics: list[str]
+
+    def output(self, results):
+        pass
+
+
+@dataclass
+class CompositeMetric(Metric):
+
+    unit_metrics: list[str] = None
+
+    def __post_init__(self):
+        if self._unit_metrics:
+            self.unit_metrics = self._unit_metrics
+        elif self.unit_metrics is None:
+            raise ValueError("unit_metrics must be provided")
+
+        if hasattr(self, "_output_template"):
+            self.output_template = self._output_template
+
+    def run(self):
+        self.result = run_metrics(
+            test_id=self.test_id,
+            metric_ids=self.unit_metrics,
+            inputs=self._get_input_dict(),
+            params=self.params,
+            output_template=self.output_template,
+            show=False,
+        )
+
+        return self.result
+
+    def summary(self, result: dict):
+        return ResultSummary(results=[ResultTable(data=[result])])
+
+
+def load_composite_metric(
+    metric_key: str = None,
+    metric_name: str = None,
+    unit_metrics: list[str] = None,
+    output_template: str = None,
+) -> CompositeMetric:
+    # this function can either create a composite metric from a list of unit metrics or
+    # load a stored composite metric based on the test id
+
+    # TODO: figure out this circular import thing:
+    from ..api_client import get_metadata
+
+    if metric_key:
+        # get the unit metric ids and output template (if any) from the metadata
+        unit_metrics = run_async(
+            get_metadata, f"composite_metric_def:{metric_key}:unit_metrics"
+        )["json"]
+        output_template = run_async(
+            get_metadata, f"composite_metric_def:{metric_key}:output_template"
+        )["json"]["output_template"]
+
+    class_def = type(
+        metric_key.split(".")[-1] if metric_key else metric_name,
+        (CompositeMetric,),
+        {
+            "__doc__": "Composite Metric built from multiple unit metrics",
+            "_unit_metrics": unit_metrics,
+            "_output_template": output_template,
+        },
+    )
+
+    required_inputs = set()
+    for metric_id in unit_metrics:
+        metric_cls = _get_metric_class(metric_id)
+        required_inputs.update(_extract_required_inputs(metric_cls))
```

Comment:
Let me know what you think of this solution @cachafla... The alternative would be to expect that unit metrics set the `required_inputs` property as part of the class definition.

Reply:
Ah... I didn't notice the first set of unit metrics don't have `required_inputs` set. I think it's fine to expect the unit metrics to define this. We can fix the existing classes to have that information.

Reply:
Sounds good